### PR TITLE
Map CAP area display options

### PR DESCRIPTION
**Pull request by biplovbhandari**
for issue https://github.com/sahana/SAMBRO/issues/74

### CODE REVIEW

**Comment by flavour** [comment]
I find this rather unclear I'm afraid. I think the deployment_setting could use a summary of what it does in the docstring...can you add that?

**Comment by biplovbhandari** [comment]
@flavour please review

File: modules/s3db/cap.py [hunk 1]
```diff
@@ -2142,36 +2151,47 @@ def cap_area_tag_onaccept(form):
...
+            # code since polygon change over time, even if the code remains
+            # the same. Hence the historic polygon is excluded as (gtable.end_date == None)
+            tquery = (ttable.tag == same_code) & \
+                     (ttable.value == form_vars.get("value")) & \
+                     (ttable.deleted != True) & \
+                     (ttable.location_id == gtable.id) & \
+                     (gtable.end_date == None) & \
+                     (gtable.deleted != True)
+            trows = db(tquery).select(ttable.location_id)
+            for trow in trows:
```
Comment (flavour) [inline modules/s3db/cap.py]: I know this is part of old code, not the change, but we would never expect to see multiple rows here would we? In which case a limitby=(0,1) will improve performance noticeably.

File: modules/s3db/cap.py [hunk 2]
```diff
@@ -2097,7 +2097,9 @@ def cap_area_onvalidation(form):
     @staticmethod
     def cap_area_location_onaccept(form):
         """
-            Link alert_id for non-template area
+            - Link alert_id for non-template area
+            - for external alerts (from import feed or rss), make sure the
+              locations are imported from the given setting
```
Comment (flavour) [inline modules/s3db/cap.py]: locations are only imported if we set polygons to be imported

File: modules/s3db/cap.py [hunk 3]
```diff
@@ -2117,7 +2119,14 @@ def cap_area_location_onaccept(form):
         if alert_id:
             # This is not template area
             # NB Template area are not linked with alert_id
-            db(db.cap_area_location.id == form_vars["id"]).update(alert_id = alert_id)
+            table = current.s3db.cap_alert
+            location_default = current.deployment_settings.get_cap_area_default()
+            row = db(table.id == alert_id).select(table.external,
+                                                  limitby=(0, 1)).first()
+            if row.external and "polygon" not in location_default:
```
Comment (flavour) [inline modules/s3db/cap.py]: The DB call here is only required if polygon not in location_default...so move it inside that branch to improve performance for the cases where polygon is in location_default

File: modules/s3db/cap.py [hunk 4]
```diff
@@ -2142,36 +2151,47 @@ def cap_area_tag_onaccept(form):
         arow = db(atable.id == area_id).select(atable.alert_id,
                                                limitby=(0, 1)).first()
         alert_id = arow.alert_id
+        table = current.s3db.cap_alert
+        location_default = current.deployment_settings.get_cap_area_default()
+        row = db(table.id == alert_id).select(table.external,
+                                              limitby=(0, 1)).first()
+        import_location = False
+        if row.external:
+            if "geocode" in location_default and tag and same_code:
```
Comment (flavour) [inline modules/s3db/cap.py]: The DB call here is only required if geocode in location_default...so move it inside that branch to improve performance for the cases where geocode is not in location_default

File: modules/s3db/cap.py (hunk 3 above)
Comment (biplovbhandari) [inline modules/s3db/cap.py]: 👍

File: modules/s3db/cap.py (hunk 4 above)
Comment (biplovbhandari) [inline modules/s3db/cap.py]: 👍

File: modules/s3db/cap.py (hunk 1 above)
Comment (biplovbhandari) [inline modules/s3db/cap.py]: 👍

**Comment by biplovbhandari** [comment]
@flavour Please merge

File: modules/s3db/cap.py [hunk 5]
```diff
@@ -2117,7 +2119,15 @@ def cap_area_location_onaccept(form):
...
             # This is not template area
             # NB Template area are not linked with alert_id
-            db(db.cap_area_location.id == form_vars["id"]).update(alert_id = alert_id)
+            location_default = current.deployment_settings.get_cap_area_default()
+            if "polygon" not in location_default:
+                table = current.s3db.cap_alert
+                row = db(table.id == alert_id).select(table.external,
+                                                      limitby=(0, 1)).first()
+                if row.external:
+                    db(db.cap_area_location.id == form_vars["id"]).delete()
```
Comment (flavour) [inline modules/s3db/cap.py]: Would be good to add a comment on why here

File: modules/s3db/cap.py [hunk 6]
```diff
@@ -2117,7 +2119,15 @@ def cap_area_location_onaccept(form):
...
+            else:
+                db(db.cap_area_location.id == form_vars["id"]).update(alert_id = alert_id)
```
Comment (flavour) [inline modules/s3db/cap.py]: More importantly, should this not run when polygon not in location_default but row.external is False?

File: modules/s3db/cap.py [hunk 7]
```diff
@@ -2142,36 +2152,47 @@ def cap_area_tag_onaccept(form):
...
+                row = db(table.id == alert_id).select(table.external,
+                                                      limitby=(0, 1)).first()
+                if row.external:
+                    import_location = True
+        else:
+            # For system alert
+            import_location = True
+
+        if import_location:
+            # SAME tag referes to some location_id in CAP
```
Comment (flavour) [inline modules/s3db/cap.py]: typo: refers

**Pull request closed without merging**